pkg/monitor/azure/nsg/nsg.go

```diff
@@ -0,0 +1,168 @@
+package nsg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"net/netip"
+	"strings"
+	"sync"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v2"
+	"github.com/sirupsen/logrus"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/metrics"
+	"github.com/Azure/ARO-RP/pkg/monitor/dimension"
+	"github.com/Azure/ARO-RP/pkg/monitor/emitter"
+)
+
+const (
+	MetricInvalidDenyRule          = "monitor.preconfigurednsg.invaliddenyrule"
+	MetricSubnetAccessResponseCode = "monitor.preconfigurednsg.subnetaccessresponsecode"
+	MetricSubnetAccessForbidden    = "monitor.preconfigurednsg.subnetaccessforbidden"
+	MetricUnsuccessfulFPCreation   = "monitor.preconfigurednsg.fpcreationunsuccessful"
+	MetricNSGMonitoringTimedOut    = "monitor.preconfigurednsg.monitoringtimedout"
+)
+
+var expandNSG = "NetworkSecurityGroup"
+
+// NSGMonitor is responsible for performing NSG rule validations when preconfiguredNSG is enabled
+type NSGMonitor struct {
+	log     *logrus.Entry
+	emitter metrics.Emitter
+	oc      *api.OpenShiftCluster
+
+	wg *sync.WaitGroup
+
+	subnetClient *armnetwork.SubnetsClient
+	dims         map[string]string
+}
+
+func NewNSGMonitor(log *logrus.Entry, oc *api.OpenShiftCluster, subscriptionID string, subnetClient *armnetwork.SubnetsClient, emitter metrics.Emitter, wg *sync.WaitGroup) *NSGMonitor {
+	return &NSGMonitor{
+		log:     log,
+		emitter: emitter,
+		oc:      oc,
+
+		subnetClient: subnetClient,
+		wg:           wg,
+
+		dims: map[string]string{
+			dimension.ResourceID:     oc.ID,
+			dimension.SubscriptionID: subscriptionID,
+			dimension.Location:       oc.Location,
+		},
+	}
+}
+
+type subnetNSGConfig struct {
+	// subnet CIDR range
+	prefix []netip.Prefix
+	// The rules from the subnet NSG
+	nsg *armnetwork.SecurityGroup
+}
+
+func (n *NSGMonitor) toSubnetConfig(ctx context.Context, subnetID string) (subnetNSGConfig, error) {
```

Comment:
My biggest question(s): 
- Will we create a monitor which alerts SRE if it's improperly configured?  -> **Answer: NO**
    - If so, how will we construct the monitor?  
- How will we alert SRE should the nsgs be misconfigured if an IcM is opened to us?  -> I don't think we need to answer this today, I think ASI is the proper path forward, but need to get there as this is relevant info. 
    - I.e. how can we determine which geneva health signals are firing against a cluster?

Reply:
Maybe just create a story for the second item.  No real action here.

Reply:
ARO-4502 has been created.